wnfs/src/private/file.rs

```diff
@@ -62,29 +108,370 @@ impl PrivateFile {
+        bare_name: &Namefilter,
+        content: Vec<u8>,
+        mut hamt: Rc<PrivateForest>,
+        store: &mut B,
+        rng: &mut R,
+    ) -> Result<(FileContent, Rc<PrivateForest>)> {
+        // TODO(appcypher): Use a better heuristic to determine when to use external storage.
+        if content.len() <= MAX_INLINE_CONTENT_SIZE {
+            return Ok((FileContent::Inline { data: content }, hamt));
+        }
+
+        let key = Key(get_random_bytes(rng));
+        let block_count = (content.len() as f64 / MAX_BLOCK_CONTENT_SIZE as f64).ceil() as usize;
+
+        for (index, label) in Self::generate_shard_labels(&key, block_count, bare_name).enumerate()
+        {
+            let start = index * MAX_BLOCK_CONTENT_SIZE;
+            let end = content.len().min((index + 1) * MAX_BLOCK_CONTENT_SIZE);
+            let slice = &content[start..end];
+
+            let enc_bytes = key.encrypt(&Key::generate_nonce(rng), slice)?;
+            let content_cid = store.put_block(enc_bytes, libipld::IpldCodec::Raw).await?;
+
+            hamt = hamt.put_encrypted(label, content_cid, store).await?;
+        }
+
+        Ok((
+            FileContent::External {
+                key,
+                block_count,
+                block_content_size: MAX_BLOCK_CONTENT_SIZE,
+            },
+            hamt,
+        ))
+    }
+
+    /// Gets the upper bound of a file content size.
+    pub(crate) fn get_content_size_upper_bound(&self) -> usize {
+        match &self.content {
+            FileContent::Inline { data } => data.len(),
+            FileContent::External {
+                block_count,
+                block_content_size,
+                ..
+            } => block_count * block_content_size,
         }
     }
 
+    /// Decrypts a block of a file's content.
+    async fn decrypt_block<B: BlockStore>(
+        key: &Key,
+        label: &Namefilter,
+        hamt: &PrivateForest,
+        store: &B,
+    ) -> Result<Vec<u8>> {
+        let label_hash = &Sha3_256::hash(&label.as_bytes());
+
+        let cids = hamt
+            .get_encrypted(label_hash, store)
+            .await?
+            .ok_or(FsError::FileShardNotFound)?;
+
+        let cid = cids
+            .iter()
+            .next()
+            .expect("Expected set with at least a Cid");
+
+        let enc_bytes = store.get_block(cid).await?;
+        let bytes = key.decrypt(&enc_bytes)?;
+
+        Ok(bytes)
+    }
+
+    /// Generates the labels for the shards of a file.
+    fn generate_shard_labels<'a>(
+        key: &'a Key,
+        mut block_count: usize,
+        bare_name: &'a Namefilter,
+    ) -> impl Iterator<Item = Namefilter> + 'a {
+        iter::from_fn(move || {
+            if block_count < 1 {
+                return None;
+            }
+
+            let label = Self::create_block_label(key, block_count, bare_name);
+            block_count -= 1;
+            Some(label)
+        })
+    }
```

Comment:
If I read this correctly, this generates the block labels in reverse, i.e. starting from `block_count` then `block_count-1` and ending with `0`.

This would mean that the first block of bytes for a file would be found at the accumulator of `parent_accumulator + key + hash(key ++ block_count)` instead of `parent_accumulator + key + hash(key ++ 0)`.

If that's the case, then this should change to match the spec.

```suggestion
    /// Generates the labels for the shards of a file.
    fn generate_shard_labels<'a>(
        key: &'a Key,
        block_count: usize,
        bare_name: &'a Namefilter,
    ) -> impl Iterator<Item = Namefilter> + 'a {
        let mut index = 0;
        iter::from_fn(move || {
            if index >= block_count {
                return None;
            }

            let label = Self::create_block_label(key, index, bare_name);
            index += 1;
            Some(label)
        })
    }
```

Reply:
You are right. Thought of it as an optimization but it breaks the spec. 
Will fix.